### PR TITLE
feat: Use tigervnc and setup a proper desktop GUI developer env

### PR DESCRIPTION
**Pull request by axonasif**
## Description

New things to address:
- Now someone can use `Open is VSCODE` button or `local-companion-app` and then connect to the vncsserver from a local vncClient like `RealVNC` or `tigervnc-viewer` ([User feedback](https://discord.com/channels/816244985187008514/939479684985552946/939549035168165978))
- Previously there was some random x11vnc server shutdowns that I and some users had experienced ([See discussion on discord](https://discord.com/channels/816244985187008514/920628884205097001/922669846557241384))
- There have been user requests of how they can use `chrome devtools and console` on their IPAD or other such devices, so we pack in the speedy `ungoogled-chromium`. (See some discussions on discord [[1](https://discord.com/channels/816244985187008514/890901203624534026/936866690518638633)] [[2](https://canary.discord.com/channels/816244985187008514/943357360435494942/943506191387074580)])
- I think setting up a minimal but complete desktop ENV for developing/using/testing desktop GUI apps makes more sense and makes it intuitive.
- Tigervnc + xfce4 setup felt more snappy even on my mobile network

## Related Issue(s)

This PR depends on #699 

I'm not sure why but recently I also found `workspace-full-vnc:latest` to be unable to launch vnc while updating `template-flutter`.
image

## How to test

You can try it here: https://gitpod.io/#https://github.com/axonasif/template-xfce4
Or you can use `axonasif/gitpod-workspace-full-vnc:latest` dockerimage in your `.gitpod.yml` or custom dockerfile.

## Release Notes

```release-note
workspace-vnc image now sets up an xfce4 DE with tigerVNC for additional local VNC client support
```

### CODE REVIEW

**Comment by princerachit** [comment]
The dependent PR has been merged https://github.com/gitpod-io/workspace-images/pull/699

**Comment by princerachit** [comment]
Build is timing out. It waits for input during installation and fails. Is this due to the recent refactoring that happened in base image?
```
#5 30.54 debconf: unable to initialize frontend: Dialog
#5 30.54 debconf: (TERM is not set, so the dialog frontend is not usable.)
#5 30.54 debconf: falling back to frontend: Readline
#5 30.69 Configuring keyboard-configuration
#5 30.69 ----------------------------------
```

**Comment by axonasif** [comment]
> Build is timing out. It waits for input during installation and fails. Is this due to the recent refactoring that happened in base image?
> 
> ```
> #5 30.54 debconf: unable to initialize frontend: Dialog
> #5 30.54 debconf: (TERM is not set, so the dialog frontend is not usable.)
> #5 30.54 debconf: falling back to frontend: Readline
> #5 30.69 Configuring keyboard-configuration
> #5 30.69 ----------------------------------
> ```

It was expected to fail. I have to re-trigger the the build since you merged the dependant PR 😄

File: chunks/tool-vnc/.xinitrc [hunk 1]
```diff
@@ -0,0 +1,56 @@
+#!/bin/bash
+#
+
+: "${DISPLAY:=":0"}"
+export DISPLAY
+
+# # Load profile
```
Comment (princerachit) [inline chunks/tool-vnc/.xinitrc:7]: Please get rid of commented code.

**Review by princerachit — DISMISSED** [review]
Is there a reference I can use to verfy the changes in this PR? I am not well versed with VNC

**Comment by axonasif** [comment]
> Is there a reference I can use to verfy the changes in this PR? I am not well versed with VNC

I'm not entirely sure how to properly answer this question but I will try 🙏 
The main points:
- Previously we used [x11vnc](https://github.com/LibVNC/x11vnc) but now we use [tigervnc](https://tigervnc.org/)
- Previously we used to setup a minimal openbox-wm environment but not we setup a complete xfce4 desktop env.
- More details in the PR description.

**Comment by axonasif** [comment]
Hey @princerachit, seems like dazzle needs to run way longer on github actions but it's getting killed due to timeout.
Screenshot 2022-03-13 at 9 09 16 PM

**Comment by princerachit** [comment]
Hey @axonasif , It seems like there is a test step which got stuck for a long time. 
image

This could have been intermittent issue so I have re triggered the build, if you see the same error again then it is likely due to the changes in the PR. Although, I am not sure why the java combination would be affected but changes in vnc

**Comment by Furisto** [comment]
@axonasif  Executing the java test commands myself in the container I see that tigerVNC is started. I guess when running the tests it gets stuck somehow and that causes the timeout. We should prevent tigerVNC from getting started in the first place in tests.

```
bash -i -c mvn -v
-v: pyenv: command not found
-v: pyenv: command not found
mount: /dev/shm: permission denied.
info[gp-vncsession]: Failed to increase shmem
info[gp-vncsession]: Starting tigerVNC server on port 5900
Killing Xtigervnc process ID 166... which was already dead
Cleaning stale pidfile '/home/gitpod/.vnc/52616bb23f61:0.pid'!
info[gp-vncsession]: Waiting for the desktop to be fully loaded ...
[ERROR] Could not create local repository at /workspace/m2-repository -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/LocalRepositoryNotAccessibleException
```

**Comment by axonasif** [comment]
Hey @Furisto, thanks a lot for looking this up 🙏 
I made a change and now it should hopefully skip in non-gitpod-session environments.

**Review by Furisto — APPROVED** [review]

**Comment by Furisto** [comment]
Looks good to me. Build runs in reasonable time and I could connect with VNC.

**Comment by Furisto** [comment]
@axonasif Oh and can you write a release note please?

**Comment by axonasif** [comment]
> @axonasif Oh and can you write a release note please?

Done 👍

**Comment by kylos101** [comment]
Wow, nice work team (@axonasif , @princerachit , @Furisto )!!!

@axonasif you are on :fire: ! Appreciate it, and I'm sure the community will too. Once this is built in main (in 1 hour), it'll be available a timestamped image on Docker Hub.

Also, the next time we ship new `workspace-clusters`, that is when this will move to the `latest` tag.

**Comment by ChristianHuff-DEV** [comment]
Now that this has been merged, will you provide an official image (`gitpod/gitpod-workspace-full-vnc:latest`) for it?

**Pull request merged**